lib/runtime/allocator/freeing_bump.go

```diff
@@ -0,0 +1,586 @@
+	f.stats.collect()
+
+	return headerPtr + HeaderSize, nil
+}
+
+// Deallocate deallocates the space which was allocated for a pointer
+//
+// The identity or the type of the passed memory object does not matter. However, the size
+// of memory cannot shrink compared to the memory passed in previous invocations.
+//
+// NOTE: Once the allocator has returned an error all subsequent requests will return an error.
+//
+// - `mem` - a slice representing the linear memory on which this allocator operates.
+// - `ptr` - pointer to the allocated chunk
+func (f *FreeingBumpHeapAllocator) Deallocate(mem runtime.Memory, ptr uint32) (err error) {
+	if f.poisoned {
+		return ErrAllocatorPoisoned
+	}
+
+	defer func() {
+		if err != nil {
+			f.poisoned = true
+		}
+	}()
+
+	if mem.Size() < f.lastObservedMemorySize {
+		return ErrMemoryShrunk
+	}
+
+	f.lastObservedMemorySize = mem.Size()
+
+	headerPtr, ok := checkedSub(ptr, HeaderSize)
+	if !ok {
+		return fmt.Errorf("%w: %d", ErrInvalidPointerForDealocation, ptr)
+	}
+
+	header, err := readHeaderFromMemory(mem, headerPtr)
+	if err != nil {
+		return fmt.Errorf("read header from memory: %w", err)
+	}
+
+	order, ok := header.intoOccupied()
+	if !ok {
+		return ErrEmptyHeader
+	}
+
+	// update the just freed header and knit it back to the free list
+	prevHeader := f.freeLists.replace(order, Ptr{headerPtr})
+	err = writeHeaderInto(Free{prevHeader}, mem, headerPtr)
+	if err != nil {
+		return fmt.Errorf("writing header into: %w", err)
+	}
+
+	newBytesAllocated, ok := checkedSub(f.stats.bytesAllocated, order.size()+HeaderSize)
+	if !ok {
+		return fmt.Errorf("underflow of the current allocated bytes count")
+	}
+
+	f.stats.bytesAllocated = newBytesAllocated
+	f.stats.collect()
+	return nil
+}
+
+func bump(bumper *uint32, size uint32, mem runtime.Memory) (uint32, error) {
+	requiredSize := uint64(*bumper) + uint64(size)
+
+	if requiredSize > uint64(mem.Size()) {
+		requiredPages, ok := pagesFromSize(requiredSize)
+		if !ok {
+			return 0, fmt.Errorf("%w: required size %d dont fit uint32",
+				ErrAllocatorOutOfSpace, requiredSize)
+		}
+
+		currentPages := mem.Size() / PageSize
+		if currentPages >= requiredPages {
+			panic(fmt.Sprintf("current pages %d >= required pages %d", currentPages, requiredPages))
+		}
+
+		if currentPages >= MaxWasmPages {
+			return 0, fmt.Errorf("%w: current pages %d greater than max wasm pages %d",
+				ErrAllocatorOutOfSpace, currentPages, MaxWasmPages)
+		}
+
+		if requiredPages > MaxWasmPages {
+			return 0, fmt.Errorf("%w: required pages %d greater than max wasm pages %d",
+				ErrAllocatorOutOfSpace, requiredPages, MaxWasmPages)
+		}
+
+		// ideally we want to double our current number of pages,
+		// as long as it's less than the double absolute max we can have
+		nextPages := min(currentPages*2, MaxWasmPages)
+		// ... but if even more pages are required then try to allocate that many
+		nextPages = max(nextPages, requiredPages)
```

Comment:
could this hit a case where required pages > MaxWasmPages? Would this be addressed by the grow call below?

Reply:
No, because on line 524 there is the following check

```go
if requiredPages > MaxWasmPages {
	return 0, fmt.Errorf("%w: required pages %d greater than max wasm pages %d",
		ErrAllocatorOutOfSpace, requiredPages, MaxWasmPages)
}
```